Widen inference candidates for error reporting

sandersn opened the pull request:
Both type inference and type inference error reporting widen candidates the same way now.

Previously error reporting hit an assert when checking for best common supertype: it hit a condition that indicated there there *was* a best common supertype, even though the error reporting should only be called when there is no best common supertype.

Both type inference and error reporting call `getInferenceCandidates`, so I just moved the additional widening code from type inference to `getInferenceCandidates`.

Fixes #15116

sandersn commented:
This PR now fixes the bad inference that led to the error in the first place. `getCommonSupertype` wasn't correctly discarding `| null | undefined` from types.

sandersn commented on src/compiler/checker.ts:
```diff
@@ -10338,17 +10345,8 @@ namespace ts {
             if (!inferredType) {
                 const inferences = getInferenceCandidates(context, index);
                 if (inferences.length) {
-                    // We widen inferred literal types if
```
This code moves into `getInferenceCandidates` so that both type inference (this section) and type inference error reporting (the other call to `getInferenceCandidates`) widen the same way.

ahejlsberg commented:
Would you mind merging master into this branch to get thing updated, then I'll take a look.

sandersn commented:
I forgot how much has changed in inference in the last few weeks! It's merged now but it needed some rewriting to work with InferenceInfo.

sandersn commented:
Travis failed because `getNullableType` doesn't handle all the cases that `includeFalsyTypes` used to -- void, in this case. I changed `getNullableType` to handle void, and added `getOptionalType` to handle the common case of adding `| undefined` to a type. (All other calls to `getNullableType` just passed `TypeFlags.Undefined`, so I changed them to call `getOptionalType`.)

sandersn commented on src/compiler/checker.ts:
```diff
@@ -9884,12 +9884,8 @@ namespace ts {
-            const supertype = getSupertypeOrUnion(primaryTypes);
-            return supertype && getNullableType(supertype, getFalsyFlagsOfTypes(types) & TypeFlags.Nullable);
+            const supertype = getSupertypeOrUnion(map(types, getNonNullableType));
+            return supertype && getNullableType(supertype, getFalsyFlagsOfTypes(types) & (TypeFlags.Nullable | TypeFlags.Void));
```
Here's what caused the addition `TypeFlags.Void` to `getFalsyFlagsOfTypes`. When `types` contains `void`, the old `filter(types ...` code did not remove it because `void` only has `TypeFlags.Void`. But `map(types, getNonNullableType)` relies on `getTypeFacts`, which treats `void` and `undefined` exactly the same. So it removes `void` and `undefined` both. This causes the following bug:

```ts
declare function f<T>(cb: () => T | undefined): T;
f(() => { });
  ~~~~~~~~~
  Argument of type '() => void' is not assignable to parameter of type '() => undefined'.
```

I would like getNonNullableType not to remove void, but I'm not sure how to do that without changing getTypeFacts, which I am not at **all** comfortable doing.
